### ISSUE TITLE
question about learning rate scheduler

### ISSUE THREAD

**Issue by dao027** [ultralytics/yolov5#11537]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

using default hyperparameters, after training 300 epochs, result is as below:

![results](https://github.com/ultralytics/yolov5/assets/35993172/b4ac4a83-51ea-426f-8a95-508cc712778e)

i think loss and AP curve hasn't been stationary, so i want to continue training, but simply changing epochs (set 600) leads to similar result. should i change the default learning rate scheduler?please tell me how OvO

**Comment by glenn-jocher** [comment]
@dao027 hello!

Glad to hear you are working with YOLOv5. 

If you are training on a new dataset or with different data augmentations, it is expected to see fluctuations in the loss and mAP curve. Since you have not achieved a stationary result, it is recommended to continue training. 

You mentioned that simply increasing the number of epochs led to similar results. In this case, the learning rate scheduler can be the cause. Changing the default learning rate scheduler can help. Some popular options include CosineAnnealingLR, OneCycleLR, and ReduceLROnPlateau. These can be implemented using PyTorch's learning rate scheduler. 

I hope this helps! Let me know if you have any other questions.

**Comment by dao027** [comment]
> @dao027 hello!
> 
> Glad to hear you are working with YOLOv5.
> 
> If you are training on a new dataset or with different data augmentations, it is expected to see fluctuations in the loss and mAP curve. Since you have not achieved a stationary result, it is recommended to continue training.
> 
> You mentioned that simply increasing the number of epochs led to similar results. In this case, the learning rate scheduler can be the cause. Changing the default learning rate scheduler can help. Some popular options include CosineAnnealingLR, OneCycleLR, and ReduceLROnPlateau. These can be implemented using PyTorch's learning rate scheduler.
> 
> I hope this helps! Let me know if you have any other questions.

can u teach me how to change the default learning rate scheduler in detail pls?
i just know i can set hyperparameter in [data/hyps/hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml) and LR decreases from 0.01 to 0.0001 after warmup
```
lr0: 0.01  # initial learning rate (SGD=1E-2, Adam=1E-3)
lrf: 0.01  # final OneCycleLR learning rate (lr0 * lrf)
```

**Comment by glenn-jocher** [comment]
@dao027 of course! In YOLOv5, you can change the default learning rate scheduler in `train.py` by modifying the `create_scheduler` function. 

Here is an example of how to implement the  CosineAnnealingLR learning rate scheduler:

1. In `train.py`, add the following import statement:

```
from torch.optim.lr_scheduler import CosineAnnealingLR
```

2. In the `create_scheduler` function, replace the default scheduler with the CosineAnnealingLR scheduler:

```
def create_scheduler(optimizer, total_epochs, batches_per_epoch, cfg):
    lf = lambda x: ((1 + math.cos(x * math.pi / total_epochs)) / 2) * (1 - cfg['lrf']) + cfg['lrf']  # cosine
    scheduler = CosineAnnealingLR(optimizer, T_max=total_epochs, eta_min=cfg['lr0'] * cfg['lrf'], last_epoch=-1)
    return scheduler, lf
```

3. Update your hyperparameters in `data/hyps/hyp.scratch-low.yaml`, specifically the `lr0` and `lrf` values, to match the requirements of the new scheduler. 

I hope this helps! Don't hesitate to reach out if you have any more questions.

**Comment by dao027** [comment]
> @dao027 of course! In YOLOv5, you can change the default learning rate scheduler in `train.py` by modifying the `create_scheduler` function.
> 
> Here is an example of how to implement the CosineAnnealingLR learning rate scheduler:
> 
> 1. In `train.py`, add the following import statement:
> 
> ```
> from torch.optim.lr_scheduler import CosineAnnealingLR
> ```
> 
> 2. In the `create_scheduler` function, replace the default scheduler with the CosineAnnealingLR scheduler:
> 
> ```
> def create_scheduler(optimizer, total_epochs, batches_per_epoch, cfg):
>     lf = lambda x: ((1 + math.cos(x * math.pi / total_epochs)) / 2) * (1 - cfg['lrf']) + cfg['lrf']  # cosine
>     scheduler = CosineAnnealingLR(optimizer, T_max=total_epochs, eta_min=cfg['lr0'] * cfg['lrf'], last_epoch=-1)
>     return scheduler, lf
> ```
> 
> 3. Update your hyperparameters in `data/hyps/hyp.scratch-low.yaml`, specifically the `lr0` and `lrf` values, to match the requirements of the new scheduler.
> 
> I hope this helps! Don't hesitate to reach out if you have any more questions.

so many thanks ! OvO

**Comment by glenn-jocher** [comment]
@dao027 you're welcome! I'm glad I could help. If you have any other questions or if there's anything else I can assist you with, don't hesitate to reach out. Best of luck with your YOLOv5 training!
